langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java

```diff
@@ -0,0 +1,120 @@
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.ballerinalang.langlib.floatingpoint;
+
+import io.ballerina.runtime.api.creators.ErrorCreator;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.internal.FloatUtils;
+import io.ballerina.runtime.internal.util.exceptions.BLangExceptionHelper;
+import io.ballerina.runtime.internal.util.exceptions.BallerinaErrorReasons;
+import io.ballerina.runtime.internal.util.exceptions.RuntimeErrors;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.text.DecimalFormat;
+
+import static io.ballerina.runtime.api.constants.RuntimeConstants.FLOAT_LANG_LIB;
+import static io.ballerina.runtime.internal.util.exceptions.BallerinaErrorReasons.getModulePrefixedReason;
+
+/**
+ * Native implementation of lang.float:toExpString(float, int?).
+ *
+ * @since 2201.1.0
+ */
+public class ToExpString {
+
+    public static BString toExpString(double x, Object fractionDigits) {
+        // If `x` is NaN or infinite, the result will be the same as `value:toString`.
+        BString str = FloatUtils.getBStringValue(x);
+        if (str != null) {
+            return StringUtils.fromString(StringUtils.getStringValue(x, null));
+        }
+
+        long noOfFractionDigits;
+
+        // If fractionalDigits is `()`, use the minimum number of digits required to accurately represent the value.
+        double xAbsValue = Math.abs(x);
+        if (fractionDigits == null) {
+            if (xAbsValue == 0) {
+                noOfFractionDigits = BigDecimal.valueOf(xAbsValue).scale();
+            } else {
+                int integerPart = (int) (Math.log10(xAbsValue));
+                noOfFractionDigits = BigDecimal.valueOf(xAbsValue / Math.pow(10, integerPart)).scale();
+            }
+        } else {
+            noOfFractionDigits = (long) fractionDigits;
+        }
+
+        // Panic if noOfFractionDigits < 0.
+        if (noOfFractionDigits < 0) {
+            throw ErrorCreator.createError(getModulePrefixedReason(FLOAT_LANG_LIB,
+                    BallerinaErrorReasons.INVALID_FRACTION_DIGITS_ERROR),
+                    BLangExceptionHelper.getErrorDetails(RuntimeErrors.INVALID_FRACTION_DIGITS));
+        }
+
+        // Handle very large int values since they might cause overflows.
+        if (noOfFractionDigits > Integer.MAX_VALUE ||
+                (noOfFractionDigits * Long.bitCount(noOfFractionDigits) > Integer.MAX_VALUE)) {
+            // Maximum value a double can represent is around 1.7*10^308.
+            noOfFractionDigits = 308;
+        }
+
+        int exponent = (int) noOfFractionDigits;
```

Comment:
Any reason to bring `noOfFractionDigits` as long and then cast to int?

Reply:
The reason is mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/35703#discussion_r846054633